src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+    auto replicate_stages = partition->replicate_in_stages(
+      op->batch_id, std::move(placeholders.batches.front()), op->opts);
+
+    fallback.cancel();
+
+    // Forward future result to the 'op'. The expectation is that at this point
+    // the target promises (inside 'op') are used to generate futures and these
+    // futures are awaited.
+    replicate_stages.request_enqueued.forward_to(
+      std::move(op->request_enqueued));
+
+    auto replicate_fut = std::move(replicate_stages.replicate_finished)
+                           .then(
+                             [](result<cluster::kafka_result> res)
+                               -> result<raft::replicate_result> {
+                                 if (res.has_error()) {
+                                     return res.error();
+                                 }
+                                 return raft::replicate_result{
+                                   model::offset(res.value().last_offset())};
+                             });
+
+    replicate_fut.forward_to(std::move(op->replicate_finished));
+}
+
+ss::future<result<model::offset>> cloud_topic_partition::replicate(
+  chunked_vector<model::record_batch> batches, raft::replicate_options opts) {
+    using ret_t = result<model::offset>;
+    chunked_vector<model::record_batch_header> headers;
+    headers.reserve(batches.size());
+    for (const auto& batch : batches) {
+        headers.push_back(batch.header());
+    }
+    // TODO: use a config
+    auto default_timeout = 1s;
+
+    // Dataplane.
+    auto res = co_await _ct_api->write_and_debounce(
+      ntp(), std::move(batches), opts.timeout.value_or(default_timeout));
+
+    if (res.has_error()) {
+        co_return ret_t(res.error());
+    }
+
+    auto placeholders = convert_to_placeholders(
+      res.value(), std::move(headers));
+
+    chunked_vector<model::record_batch> placeholder_batches;
+    for (auto&& batch : placeholders.batches) {
+        placeholder_batches.push_back(std::move(batch));
+    }
+
+    // Control plane.
+    auto result = co_await _partition->replicate(
+      std::move(placeholder_batches), opts);
+
+    if (!result) {
+        co_return ret_t(result.error());
+    }
+    co_return ret_t(model::offset(result.value().last_offset()));
+}
+
+ss::future<result<model::offset>> cloud_topic_partition::replicate(
+  model::record_batch batch, raft::replicate_options opts) {
+    return replicate(
+      chunked_vector<model::record_batch>::single(std::move(batch)), opts);
+}
+
+raft::replicate_stages cloud_topic_partition::replicate(
+  model::batch_identity batch_id,
+  model::record_batch batch,
+  raft::replicate_options opts) {
+    auto header = batch.header();
+    chunked_vector<model::record_batch> batch_vec;
+    batch_vec.push_back(std::move(batch));
+    auto op_state = ss::make_lw_shared<upload_and_replicate_stages>(
+      _partition,
+      std::move(batch_vec),
+      batch_id,
+      opts,
+      opts.timeout.value_or(1s));
```

Comment:
another hardcoded `1s`, can we at least for the time being pull them up into some `static constexpr const` value they all share?